[mlir][linalg] Enable masked vectorisation for depthwise convolutions

banach-space opened the pull request:
This patch adds support for masked vectorisation of depthwise 1D WC
convolutions,`linalg.depthwise_conv_1d_nwc_wc`. This is implemented by
adding support for masking.

Two major assumptions are made:
  * only the channel dimension can be dynamic/scalable (i.e. the
    trailing dim),
  * when specifying vector sizes to use in the vectoriser, only the size
    corresponding to the channel dim is effectively used (other dims are
    inferred from the context).

In terms of scalable vectorisation, this should be sufficient to cover
all practical cases (i.e. making arbitrary dim scalable wouldn't make
much sense). As for more generic cases with dynamic shapes (e.g. W or N
dims being dynamic), more work would be needed. In particular, one would
have to consider the filter and input/output tensors separately.

banach-space committed:
[mlir][linalg] Add scalable vectorisation for depthwise convolutions

This patch adds support for scalable vectorisation of depthwise 1D HWC
convolutions,`linalg.depthwise_conv_1d_nwc_wc`. This is implemented by
adding support for masking.

Two major assumptions are made:
  * only the channel dimension can be scalable/dynamic (i.e. the
    trailing dim),
  * when specifying vector sizes to use in the vectoriser, only the size
    corresponding to the channel dim is effectively used (other dims are
    inferred from the context).

In terms of scalable vectorisation, this should be sufficient that cover
all practical cases (i.e. making arbitrary dim scalable wouldn't make
much sense). As for more generic cases with dynamic shapes (e.g. w or n
dims being dynamic), more work would be needed. In particular, one would
have to consider the filter and input/ouput tensors separately. However,
it's not clear whether that would be of any use in practice.

c-rhodes commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3127,10 +3207,12 @@ struct Conv1DGenerator
+        loc, maybeMaskedRES->getResult(0), resShaped,
+        ValueRange{zero, zero, zero});
+    return maybeMaskXferOp(resType.getShape(),
+                           /*scalableDims=*/{false, false, scalableChDim},
```
```suggestion
                          resType.getScalableDims(),
```

c-rhodes commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3171,8 +3253,9 @@ struct Conv1DGenerator
       return nullptr;
 
-    if (isa<FloatType>(resTy.getElementType()))
+    if (isa<FloatType>(resTy.getElementType())) {
       return rewriter.create<vector::FMAOp>(loc, lhs, rhs, res);
+    }
```
unrelated change

c-rhodes commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3027,20 +3051,74 @@ struct Conv1DGenerator
+
+    // Masks the input xfer Op along the channel dim, iff the corresponding
+    // scalable flag is set.
+    auto maybeMaskXferOp = [&](ArrayRef<int64_t> maskShape,
```
this seems pretty generic, is it worth moving to a helper?

banach-space replied:
I'm about to update it in a way that's going to make it less generic :)

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,161 @@
+// CHECK-SAME:      %[[FILTER:.*]]: tensor<1x?xi8>,
+// CHECK-SAME:      %[[OUTPUT:.*]]: tensor<1x8x?xi8>) -> tensor<1x8x?xi8> {
+
+// CHECK-DAG:       arith.constant 1 : index
```
can be removed

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,161 @@
+
+// CHECK-DAG:       arith.constant 1 : index
+// CHECK-DAG:       %[[VAL_5:.*]] = arith.constant 1 : index
+// CHECK-DAG:       %[[VAL_6:.*]] = tensor.dim %[[FILTER]], %[[VAL_5]] : tensor<1x?xi8>
```
```suggestion
// CHECK-DAG:       %[[CH_DIM:.*]] = tensor.dim %[[FILTER]], %[[VAL_5]] : tensor<1x?xi8>
```

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,161 @@
+// CHECK-DAG:       %[[VAL_5:.*]] = arith.constant 1 : index
+// CHECK-DAG:       %[[VAL_6:.*]] = tensor.dim %[[FILTER]], %[[VAL_5]] : tensor<1x?xi8>
+// CHECK-DAG:       %[[C0:.*]] = arith.constant 0 : index
+// CHECK-DAG:       %[[C0_I8:.*]] = arith.constant 0 : i8
```
```suggestion
// CHECK-DAG:       %[[PAD:.*]] = arith.constant 0 : i8
```
?

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,161 @@
+// CHECK:           %[[VEC_IN:.*]] = vector.mask %[[MASK_IN]] { vector.transfer_read %[[INPUT]]{{\[}}%[[C0]], %[[C0]], %[[C0]]], %[[C0_I8]] : tensor<1x8x?xi8>, vector<1x8x[4]xi8> } : vector<1x8x[4]xi1> -> vector<1x8x[4]xi8>
+
+/// Create a mask for the filter tensor
+// CHECK:           %[[C0_I8_1:.*]] = arith.constant 0 : i8
```
run cse? So much duplication

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,161 @@
+// CHECK:           %[[ADDI:.*]] = arith.addi %[[MULI]], %[[VEC_OUT_0]] : vector<1x8x[4]xi8>
+// CHECK:           %[[VEC_OUT_1:.*]] = vector.insert_strided_slice %[[ADDI]], %[[VEC_OUT]] {offsets = [0, 0, 0], strides = [1, 1, 1]} : vector<1x8x[4]xi8> into vector<1x8x[4]xi8>
+
+/// Create a mask for the output tensor
+// CHECK:           %[[VAL_36:.*]] = arith.constant 2 : index
+// CHECK:           %[[VAL_37:.*]] = tensor.dim %[[OUTPUT]], %[[VAL_36]] : tensor<1x8x?xi8>
+// CHECK:           %[[VAL_38:.*]] = arith.constant 1 : index
+// CHECK:           %[[VAL_39:.*]] = arith.constant 8 : index
+// CHECK:           %[[MASK_OUT:.*]] = vector.create_mask %[[VAL_38]], %[[VAL_39]], %[[VAL_37]] : vector<1x8x[4]xi1>
```
this is the same mask from earlier

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1609,6 +1610,19 @@ static LogicalResult reductionPreconditions(LinalgOp op) {
+  // Support dynamic shapes in 1D depthwise convolution, but only in the
+  // _channel_ dimension. That's exclusively to support scalable vectorisation.
+  if (auto conv = dyn_cast<linalg::DepthwiseConv1DNwcWcOp>(op.getOperation())) {
+    auto lhsShaped = op.getDpsInputOperand(0)->get();
```
Just sticking with convention used for convs in this file: https://github.com/llvm/llvm-project/blob/55a7ff8c1ac23a78f9701d357d292e6a0bdfdb4f/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L2662-L2664

c-rhodes replied:
ok, I'd drop it so it's clearly distinguishable from `lhsShape`, the fact it's a ShapedType can also be easily inferred from the non-checking cast below.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1609,6 +1610,19 @@ static LogicalResult reductionPreconditions(LinalgOp op) {
+    auto lhsShaped = op.getDpsInputOperand(0)->get();
+    ArrayRef<int64_t> lhsShape =
+        dyn_cast<ShapedType>(lhsShaped.getType()).getShape();
+    auto shapeWithoutCh = lhsShape.drop_back(1);
+    if (ShapedType::isDynamicShape(shapeWithoutCh))
+      return failure();
+
+    return success();
```
Not sure I follow - we need to verify that all dimensions except for the channel dim are static. Checking the channel dim alone wouldn't be sufficient. Perhaps share a snippet so that I know what you meant?

c-rhodes replied:
ah ok, I was suggesting
```suggestion
    if (ShapedType::isDynamicShape(lhsShape.back()))
      return success();

    return failure();
```
but as you say that's not checking if the other dims are static. Perhaps worth adding a negative test with dynamic dims other than channel to check nothing happens.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1609,6 +1610,19 @@ static LogicalResult reductionPreconditions(LinalgOp op) {
 }
 
 static LogicalResult vectorizeDynamicLinalgOpPrecondition(linalg::LinalgOp op) {
+  // Support dynamic shapes in 1D depthwise convolution, but only in the
+  // _channel_ dimension. That's exclusively to support scalable vectorisation.
```
Nothing :) This pre-condition will fail and the vectoriser won't do anything. And there are other pre-conditions too.

In terms of what's missing in the vectoriser - we'd need forward the corresponding vector sizes to `depthwiseConv`. Otherwise we'd be trying to create a vector with one dynamic shape. That's missing as I only added logic for channel dim. I am reluctant to extend that - we'd need to worry about the width dimension for the input, output and filter separately. I'm not sure whether that would be needed in practice.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3107,6 +3186,7 @@ struct Conv1DGenerator
     // Its possible we failed to create the Fma.
     if (!llvm::all_of(resVals, [](Value v) { return v; })) {
       // Manually revert (in reverse order) to avoid leaving a bad IR state.
+      // TODO: Replace with maybeMasked
```
I honestly don't remember what this comment was for - it's a leftover from experimenting, so will just remove it.

banach-space committed:
fixup! [clang][Driver] Small correction to print-runtime-dir

Addressing Cullen's comments

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1609,6 +1610,20 @@ static LogicalResult reductionPreconditions(LinalgOp op) {
 }
 
 static LogicalResult vectorizeDynamicLinalgOpPrecondition(linalg::LinalgOp op) {
+  if (auto conv = dyn_cast<linalg::DepthwiseConv1DNwcWcOp>(op.getOperation())) {
```
could we move all this conv specific logic to a specific `*Precondition` function?

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2697,6 +2714,7 @@ struct Conv1DGenerator
         return;
       break;
     }
+    hasTensorSemantics = linalgOp.hasPureTensorSemantics();
```
?

MacDue replied:
You have `op` in that function, is that the linalg op? If so you could do `cast<LinalgOp>(op).hasPureTensorSemantics()` within `maybeMaskXferOp`, rather than add a new member variable.

banach-space replied:
Thanks for the suggestion, Ben! 🙏🏻

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3000,13 +3018,21 @@ struct Conv1DGenerator
     if (!valid)
       return rewriter.notifyMatchFailure(op, "unvectorizable depthwise conv");
 
+    bool scalableChDim = false;
     int64_t nSize, wSize, cSize, kwSize;
     // kernel{kw, c}
     bindShapeDims(rhsShapedType, kwSize, cSize);
+    // Dynamic channel size implies scalable vectorisation
+    if (ShapedType::isDynamic(cSize)) {
+      assert(channelDimVecSize != 0 && "Channel dim vec size must be > 0");
+      cSize = channelDimVecSize;
+      scalableChDim = true;
+    }
```
Why a dynamic channel dimension implies scalable vectors? We have to make sure this also works for non scalable cases.

MacDue replied:
I wonder this too (though I'm not very familiar with the linalg vectorizer). Could also use scalable vectors if your channel dimension was large (say 100 elements), but still a static size?

Would it be possible to control this with the by passing a the scalable dims flags for the vector sizes, like is the case with matmuls?

banach-space replied:
> Why a dynamic channel dimension implies scalable vectors?

Because I run out of time this week an that's the case I need the most :) I will send an update shortly to add support for more generic case too. It's actually quite easy. The most labor-intensive task is writing tests for these convs.

> Could you also use scalable vectors if your channel dimension was large (say 100 elements), but still a static size?

Yes, we already do that for elementwise Ops. However, you need to tile first - that will lead to tensors with dynamic shapes.

dcaballe reviewed (commented):
Thanks! Taking a first look!

c-rhodes commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1609,6 +1610,19 @@ static LogicalResult reductionPreconditions(LinalgOp op) {
+        dyn_cast<ShapedType>(lhsShaped.getType()).getShape();
+    auto shapeWithoutCh = lhsShape.drop_back(1);
+    if (ShapedType::isDynamicShape(shapeWithoutCh))
+      return failure();
```
perhaps use debug then like the other static hooks
```suggestion
      LDBG("Dynamically-shaped op vectorization precondition failed: only channel dim can be dynamic\n");
      return failure();
```

banach-space reviewed (commented):
Thanks for the reviews, sending updates shortly!

banach-space committed:
fixup! [mlir][linalg] Add scalable vectorisation for depthwise convolutions

Addressing PR comments:
- add CSE in tests, update check-lines accordingly
- add support for plain (non-scalable) masked vectorisation
- moved pre-conditions for vectorisation to a dedicated hook

cfRod commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+
+// -----
+
+func.func @depthwise_conv1d_nwc_wc_3x5x4xf32_memref_dillation_2(%input: memref<3x5x?xf32>,
```
dilation? typo?

banach-space replied:
Nice catch, thanks!

cfRod commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+  return
+}
+
+// TODO - nice variable names
```
remove TODO?

MacDue commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3268,7 +3358,9 @@ struct Conv1DGenerator
   ///   {{n, strideW * w + dilationW * kw, c}, {kw, c}, {n, w, c}}
-  FailureOr<Operation *> generateDilatedConv(bool flatten = false) {
+  FailureOr<Operation *> generateDilatedConv(uint64_t vecChDimSize = 0,
+                                             bool vecChDimScalableFlag = true,
```
Should this default to true? I see it's controlled by the vector `inputScalableVecDims` now, but making it the default here is a little odd.

banach-space replied:
Leftover from experiments, thanks for pointing out!

MacDue commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3000,13 +3030,23 @@ struct Conv1DGenerator
+    if (ShapedType::isDynamic(cSize)) {
+      assert(channelDimVecSize != 0 && "Channel dim vec size must be > 0");
+      cSize = channelDimVecSize;
+      scalableChDim = channelDimScalableFlag;
```
Maybe add a note here that scalable vectors are only used if the channel dim is dynamic and `channelDimScalableFlag` is set.

banach-space committed:
fixup! fixup! [mlir][linalg] Add masked vectorisation for depthwise convolutions

Address comments from Ben and Crefeda

banach-space committed:
fixup! [mlir][linalg] Add masked vectorisation for depthwise convolutions

Better LIT var names in tests

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+// CHECK:           %[[OUT_INS:.*]] = vector.insert_strided_slice %[[FMA_2]], %[[VEC_OUT]] {offsets = [0, 0, 0], strides = [1, 1, 1]} : vector<3x2x[4]xf32> into vector<3x2x[4]xf32>
+// CHECK:           vector.mask %[[MASK_OUT]] { vector.transfer_write %[[OUT_INS]], %[[OUTPUT]]{{\[}}%[[C0]], %[[C0]], %[[C0]]] : vector<3x2x[4]xf32>, memref<3x2x?xf32> } : vector<3x2x[4]xi1>
+
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg0: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.depthwise_conv_1d_nwc_wc"]} in %arg0 : (!transform.any_op) -> !transform.any_op
+    transform.structured.vectorize %0 vector_sizes [3, 2, [4], 2] : !transform.any_op
+    transform.yield
+  }
```
move this above the CHECK lines for consistency with the other tests

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+      %filter: memref<2x?xf32>,
+      %output: memref<3x2x?xf32>) {
+  linalg.depthwise_conv_1d_nwc_wc
+    {dilations = dense<2> : tensor<1xi64>, strides = dense<1> : tensor<1xi64>}
```
what's the significance of a dilation of 2? Is dilation explained anywhere? I can't see anything in the linalg docs.

banach-space replied:
I've been struggling to find good documentation for this. This book provides a nice description:
* https://udlbook.github.io/udlbook/

See Figure 10.3.

The reason for adding this case here is to demonstrate that this patch is not doing anything that would only work for the default case (dilation = 1). But I appreciate that it's a bit tricky to follow without knowing the semantics of the Op itself. Is there anything I can do to make it clearer? You could try comparing against: https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/test/Dialect/Linalg/vectorize-convolution-flatten.mlir#L51-L108

But note that I wasn't trying to keep the check-lines consistent and, more importantly, that is testing "flattened" convs.

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+// RUN: mlir-opt -split-input-file -transform-interpreter -cse %s | FileCheck %s
+
+func.func @depthwise_conv1d_nwc_wc_1x8x3xi8_tensor(%input: tensor<1x8x?xi8>,
+                                                   %filter: tensor<1x?xi8>,
+                                                   %output: tensor<1x8x?xi8>) -> (tensor<1x8x?xi8>) {
+  %res = linalg.depthwise_conv_1d_nwc_wc
+    {dilations = dense<1> : vector<1xi64>,
+    strides = dense<1> : vector<1xi64>}
+    ins(%input, %filter : tensor<1x8x?xi8>, tensor<1x?xi8>)
+    outs(%output : tensor<1x8x?xi8>) -> tensor<1x8x?xi8>
+  return %res : tensor<1x8x?xi8>
+}
+
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg0: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.depthwise_conv_1d_nwc_wc"]} in %arg0 : (!transform.any_op) -> !transform.any_op
+    transform.structured.vectorize %0 vector_sizes [1, 8, 4, 1] : !transform.any_op
+    transform.yield
+  }
+}
+
```
this isn't scalable as the filename suggests

banach-space replied:
Good catch - I've recently added this following Diego's request to make this work for plain masking. Will rename the file accordingly.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3371,7 +3465,17 @@ vectorizeConvolution(RewriterBase &rewriter, LinalgOp op,
+  if (!inputVecSizes.empty()) {
+    // Only use the input vector size corresponding to the channel dim. Other
+    // vector dims will be inferred from the Ops.
+    vecChDimSize = inputVecSizes[2];
+    vecChDimScalableFlag = inputScalableVecDims[2];
```
Yes. Note that I am inserting this just before calling `generateDilatedConv` and: https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L3400-L3416

Specifically:
```cpp:
      return depthwiseConv(flatten);
```
and
https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L3125-L3134

I can add a comment to clarify the assumptions made.

c-rhodes replied:
I think an assert that checks this is a `depthwise_conv_1d_nwc_wc` would be more appropriate?

banach-space replied:
I've double check and at this point we hit either:
* `depthwise_conv_1d_nwc_wc`, or
* `depthwise_conv_1d_ncw_cw`.

The latter is not vectorised atm and I'll add a case to vectorization-unsupported.mlir to document that. I will also add an assert and update this code to work for both types of convs (even though NCW is not really supported further down). Thanks for the suggestion! 👍🏻

banach-space committed:
fixup! [mlir][linalg] Add masked vectorisation for depthwise convolutions

More documentaiton, some simplification (as per Cullen's comments)

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3027,20 +3069,68 @@ struct Conv1DGenerator
+        return opToMask;
+      auto maskType =
+          VectorType::get(maskShape, rewriter.getI1Type(), scalableDims);
+      SmallVector<OpFoldResult> mixedSourceDims =
+          cast<LinalgOp>(op).hasPureTensorSemantics()
+              ? TypeSwitch<Operation *, SmallVector<OpFoldResult>>(opToMask)
+                    .Case<vector::TransferReadOp>([&](auto readOp) {
+                      return tensor::getMixedSizes(rewriter, loc,
+                                                   readOp.getSource());
+                    })
+                    .Case<vector::TransferWriteOp>([&](auto writeOp) {
+                      return tensor::getMixedSizes(rewriter, loc,
+                                                   writeOp.getOperand(1));
+                    })
+              : TypeSwitch<Operation *, SmallVector<OpFoldResult>>(opToMask)
+                    .Case<vector::TransferReadOp>([&](auto readOp) {
+                      return memref::getMixedSizes(rewriter, loc,
+                                                   readOp.getSource());
+                    })
+                    .Case<vector::TransferWriteOp>([&](auto writeOp) {
+                      return memref::getMixedSizes(rewriter, loc,
+                                                   writeOp.getOperand(1));
+                    });
```
Move to utility? Couldn't this getMixedSizes be implemented on the common ShapedType?

banach-space replied:
> Move to utility? 

👍🏻 

> Couldn't this getMixedSizes be implemented on the common ShapedType?

Not really as these methods create e.g. `tensor::DimOp`/`memref::DimOp`. But yeah, otherwise these are identical.

banach-space committed:
fixup! [mlir][linalg] Add masked vectorisation for depthwise convolutions

Address Diego's comments, move  to vector utils

banach-space committed:
fixup! [mlir][linalg] Add scalable vectorisation for depthwise convolutions

Generalise the code a tiny bit to cover for 1D depthwise NCW convs (once
supported by the vectoriser).

banach-space commented:
Kind ping :)

I believe that I've addressed all your comments - could you take another look? This is quite critical for me  🙏🏻 😅

c-rhodes commented on mlir/test/Dialect/Linalg/vectorize-conv-masked-and-scalable.mlir:
```diff
@@ -0,0 +1,187 @@
+// CHECK:           %[[VEC_OUT:.*]] = vector.mask %[[MASK_OUT]] { vector.transfer_read %[[OUTPUT]]{{\[}}%[[C0]], %[[C0]], %[[C0]]], %[[PAD]] : tensor<1x8x?xi8>, vector<1x8x4xi8> } : vector<1x8x4xi1> -> vector<1x8x4xi8>
+
+/// Convolution
+// CHECK:           %[[IN_1:.*]] = vector.extract_strided_slice %[[VEC_IN]] {offsets = [0, 0, 0], sizes = [1, 8, 4], strides = [1, 1, 1]} : vector<1x8x4xi8> to vector<1x8x4xi8>
```
the insert/extract strided slice operations in this test and the scalable equivalent are not necessary, with canonicalization these are removed which I think makes it easier to follow, as well as making it more obvious what is happening for the final test where the dilation is 2 and the `extract_strided_slice` is necessary.

c-rhodes reviewed (approved):
I've left a few final comments but otherwise this LGTM, cheers. I would appreciate an integration test as well, but that doesn't need to be in this patch.

MacDue commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -3186,20 +3259,21 @@ struct Conv1DGenerator
     for (int64_t w = 0; w < wSize; w += wSizeStep) {
       resVals.push_back(rewriter.create<vector::ExtractStridedSliceOp>(
-          loc, res,
+          loc, maybeMaskedRes->getResult(0),
```
The logic below this (for when `flatten=true`) looks like it will drop the scalable dims: 
```
    auto inOutFlattenSliceSizes =
        SmallVector<int64_t>{nSize, wSizeStep * cSize};
    auto lhsCastType = VectorType::get(inOutFlattenSliceSizes, lhsEltType);
    auto resCastType = VectorType::get(inOutFlattenSliceSizes, resEltType);
    // Compute contraction: O{n, w, c} += I{n, sw * w + dw * kw, c} * F{c}
```

You may want to update that (or bail out if early if `flatten=true`).

banach-space replied:
That's a great catch, thanks! This should be captured by one of the pre-conditions and that's what I'll do.

We don't have a mechanism within TD to trigger this case, which means it's hard to test it. And I'm not sure whether extending TD just to test things is the right thing 🤔  I need to think of a use-case other than testing 😅

banach-space committed:
fixup! [mlir][linalg] Add scalable vectorisation for depthwise convolutions

* Add missing dyn dimension in a test
* Make sure "flattening" + "masked vectorisation" are not allowed

dcaballe reviewed (approved):
LGTM, for enablement purposes. I understand the criticality of this and I don't want to block the development but, as discussed (also offline with Nicolas), we know that this is a short-term specialization and that all the convolution support in the vectorizer needs to be revisited. As such, it may happen that some of this functionality has to be disabled/revisited in the future if it hinders a more generic path forward. I would also be concerned if there are plans to continue extending this path so let's make sure we discuss those and plan for a better approach if that is the case.

banach-space commented:
Thanks all for reviewing, I will be merging this shortly (it's quite critical for us - we can't leverage wider SVE vectors otherwise).

> (...)  we know that this is a short-term specialization and that all the convolution support in the vectorizer needs to be revisited.

I would be very keen to participate in any such refactor.  Lets continue brainstorming about this!

>  As such, it may happen that some of this functionality has to be disabled/revisited in the future if it hinders a more generic path forward.

I am more than happy to redesign/refactor this if need be 👍🏻
